File: services/event-streaming-service/src/models/schema_history.rs
//! Avro schema versioning (Bloque E2).

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::{FromRow, types::Json as SqlJson};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StreamSchemaVersion {
    pub id: Uuid,
    pub stream_id: Uuid,
    pub version: i32,
    pub schema_avro: Value,
    pub fingerprint: String,
    pub compatibility: String,
    pub created_by: String,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ValidateSchemaRequest {
    pub schema_avro: Value,
    /// Optional payload sample. When present it is validated against
    /// `schema_avro` and any compatibility issues against the current
    /// stream schema are also returned.
    #[serde(default)]
    pub sample: Option<Value>,
    /// Compatibility mode (Confluent style). Defaults to `BACKWARD`.
    #[serde(default)]
    pub compatibility: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct ValidateSchemaResponse {
    pub valid: bool,
    pub fingerprint: Option<String>,
    pub errors: Vec<String>,
    pub warnings: Vec<String>,
    /// When the request asked for a compatibility check, this is the
    /// outcome against the *current* persisted schema.
    pub compatibility: Option<CompatibilityOutcome>,
}

#[derive(Debug, Clone, Serialize)]
pub struct CompatibilityOutcome {
    pub mode: String,
    pub compatible: bool,
    pub reason: Option<String>,
}

#[derive(Debug, Clone, FromRow)]
pub struct StreamSchemaHistoryRow {
    pub id: Uuid,
    pub stream_id: Uuid,
    pub version: i32,
    pub schema_avro: SqlJson<Value>,
    pub fingerprint: String,
    pub compatibility: String,
    pub created_by: String,
    pub created_at: DateTime<Utc>,
}

impl From<StreamSchemaHistoryRow> for StreamSchemaVersion {
    fn from(value: StreamSchemaHistoryRow) -> Self {
        Self {
            id: value.id,
            stream_id: value.stream_id,
            version: value.version,
            schema_avro: value.schema_avro.0,
            fingerprint: value.fingerprint,
            compatibility: value.compatibility,
            created_by: value.created_by,
            created_at: value.created_at,
        }
    }
}

File: services/ingestion-replication-service/internal/models/schema_history.go
package models

// schema_history.go ports
// services/ingestion-replication-service/src/event_streaming/models/schema_history.rs.
//
// Wire types for the schema-validation + history endpoints (IRF-9):
//   POST /api/v1/streaming/streams/{id}/schema:validate
//   GET  /api/v1/streaming/streams/{id}/schema/history

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// StreamSchemaVersion is one accepted schema for a stream, surfaced
// by the history endpoint. Mirrors the Rust struct of the same name.
type StreamSchemaVersion struct {
	ID            uuid.UUID       `json:"id"`
	StreamID      uuid.UUID       `json:"stream_id"`
	Version       int32           `json:"version"`
	SchemaAvro    json.RawMessage `json:"schema_avro"`
	Fingerprint   string          `json:"fingerprint"`
	Compatibility string          `json:"compatibility"`
	CreatedBy     string          `json:"created_by"`
	CreatedAt     time.Time       `json:"created_at"`
}

// ValidateSchemaRequest is the body of POST /streams/{id}/schema:validate.
//
// Sample is optional — when present it is validated against schema_avro
// and any compatibility issues against the current stream schema are
// also returned. Compatibility, when omitted, defaults to the stream's
// persisted compatibility mode.
type ValidateSchemaRequest struct {
	SchemaAvro    json.RawMessage `json:"schema_avro"`
	Sample        json.RawMessage `json:"sample,omitempty"`
	Compatibility *string         `json:"compatibility,omitempty"`
}

// ValidateSchemaResponse is the response body. Errors and Warnings are
// always non-nil JSON arrays so the wire shape matches the Rust impl
// (which serialises empty Vecs as `[]`, not `null`).
type ValidateSchemaResponse struct {
	Valid         bool                  `json:"valid"`
	Fingerprint   *string               `json:"fingerprint"`
	Errors        []string              `json:"errors"`
	Warnings      []string              `json:"warnings"`
	Compatibility *CompatibilityOutcome `json:"compatibility"`
}

// CompatibilityOutcome reports the result of the compatibility check
// against the current persisted schema.
type CompatibilityOutcome struct {
	Mode       string  `json:"mode"`
	Compatible bool    `json:"compatible"`
	Reason     *string `json:"reason"`
}
